pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpartition/SegmentPartitionMetadataManager.java

```diff
@@ -0,0 +1,189 @@
+        Integer partitionId = segmentPartitionInfo.getPartitions().iterator().next();
+        _segmentToPartitionMap.put(segment, partitionId);
+      }
+      // update segment to server list.
+      updateSegmentServerOnlineMap(externalView, segment);
+    }
+    computePartitionMaps();
+  }
+
+  private void updateSegmentServerOnlineMap(ExternalView externalView, String segment) {
+    Map<String, String> instanceStateMap = externalView.getStateMap(segment);
+    List<String> serverList = instanceStateMap == null ? Collections.emptyList()
+        : instanceStateMap.entrySet().stream()
+            .filter(entry -> CommonConstants.Helix.StateModel.SegmentStateModel.ONLINE.equals(entry.getValue()))
+            .map(Map.Entry::getKey)
+            .collect(Collectors.toList());
+    _segmentToOnlineServersMap.put(segment, serverList);
+  }
+
+  private boolean validate(SegmentPartitionInfo segmentPartitionInfo) {
+    if (segmentPartitionInfo == null || segmentPartitionInfo.getPartitionFunction() == null
+        || segmentPartitionInfo.getPartitions() == null) {
+      return false;
+    }
+    // TODO: check more than just function name here but also function config.
+    return _partitionFunctionName.equals(segmentPartitionInfo.getPartitionFunction().getName())
+        && _partitionColumn.equals(segmentPartitionInfo.getPartitionColumn())
+        && _numPartitions == segmentPartitionInfo.getNumPartitions()
+        && segmentPartitionInfo.getPartitions().size() == 1;
+  }
+
+  @Override
+  public synchronized void onAssignmentChange(IdealState idealState, ExternalView externalView,
+      Set<String> onlineSegments, List<String> pulledSegments, List<ZNRecord> znRecords) {
+    // update segment zk metadata for the pulled segments
+    for (int idx = 0; idx < pulledSegments.size(); idx++) {
+      String segment = pulledSegments.get(idx);
+      ZNRecord znRecord = znRecords.get(idx);
+      SegmentPartitionInfo segmentPartitionInfo = SegmentPartitionUtils.extractPartitionInfoFromSegmentZKMetadata(
+          _tableNameWithType, _partitionColumn, segment, znRecord);
+      if (validate(segmentPartitionInfo)) {
+        // update segment to partition map
+        Integer partitionId = segmentPartitionInfo.getPartitions().iterator().next();
+        _segmentToPartitionMap.put(segment, partitionId);
+      } else {
+        // remove the segment from the partition map
+        _segmentToPartitionMap.remove(segment);
+      }
+    }
+    // update the server online information based on external view.
+    for (String onlineSegment : onlineSegments) {
+      // update segment to server list.
+      updateSegmentServerOnlineMap(externalView, onlineSegment);
+    }
+    _segmentToPartitionMap.keySet().retainAll(onlineSegments);
+    _segmentToOnlineServersMap.keySet().retainAll(onlineSegments);
+    computePartitionMaps();
+  }
+
+  @Override
+  public synchronized void refreshSegment(String segment, @Nullable ZNRecord znRecord) {
```

Comment:
review note: when refreshSegment, IS and EV are not available. assuming no change here thus not updating the segment to server map